internal/jobobject/jobobject.go

```diff
@@ -0,0 +1,414 @@
+	"unsafe"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"golang.org/x/sys/windows"
+)
+
+// This file provides higher level constructs for the win32 job object API.
+// Most of the core creation and management functions are already present in "golang.org/x/sys/windows"
+// (CreateJobObject, AssignProcessToJobObject, etc.) as well as most of the limit information
+// structs and associated limit flags. Whatever is not present from the job object API
+// in golang.org/x/sys/windows is located in /internal/winapi.
+//
+// https://docs.microsoft.com/en-us/windows/win32/procthread/job-objects
+
+// JobObject is a high level wrapper around a Windows job object. Holds a handle to
+// the job, a channel to receive iocp notifications about the lifecycle
+// of the job and a mutex for synchronized handle access.
+type JobObject struct {
+	handle     windows.Handle
+	mq         *messageQueue
+	handleLock sync.RWMutex
+}
+
+// JobLimits represents the resource constraints that can be applied to a job object.
+type JobLimits struct {
+	CPULimit           uint32
+	CPUWeight          uint32
+	MemoryLimitInBytes uintptr
+	MaxIOPS            int64
+	MaxBandwidth       int64
+}
+
+// Processor resource controls
+const (
+	ProcessorLimitMin  = 1
+	ProcessorLimitMax  = 10000
+	ProcessorWeightMin = 1
+	ProcessorWeightMax = 9
+)
+
+var (
+	ErrAlreadyClosed = errors.New("the handle has already been closed")
+	ErrNotRegistered = errors.New("job is not registered to receive notifications")
+)
+
+// SetResourceLimits sets resource limits on the job object (cpu, memory, storage).
+func (job *JobObject) SetResourceLimits(ctx context.Context, limits *JobLimits) error {
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	if job.handle == 0 {
+		return ErrAlreadyClosed
+	}
+
+	// Go through and check what limits were specified and apply them to the job.
+	if limits.MemoryLimitInBytes != 0 {
+		var (
+			basicLimitFlags uint32
+			eliInfo         windows.JOBOBJECT_EXTENDED_LIMIT_INFORMATION
+		)
+		basicLimitFlags |= windows.JOB_OBJECT_LIMIT_JOB_MEMORY
+		eliInfo.JobMemoryLimit = limits.MemoryLimitInBytes
+		eliInfo.BasicLimitInformation.LimitFlags = basicLimitFlags
+		_, err := windows.SetInformationJobObject(job.handle, windows.JobObjectExtendedLimitInformation, uintptr(unsafe.Pointer(&eliInfo)), uint32(unsafe.Sizeof(eliInfo)))
+		if err != nil {
+			return fmt.Errorf("failed to set extended limit info on job object: %s", err)
+		}
+	}
+
+	if limits.CPULimit != 0 || limits.CPUWeight != 0 {
+		cpuInfo := winapi.JOBOBJECT_CPU_RATE_CONTROL_INFORMATION{}
+		if limits.CPUWeight != 0 {
+			if limits.CPUWeight < ProcessorWeightMin || limits.CPUWeight > ProcessorWeightMax {
+				return fmt.Errorf("processor weight value of `%d` is invalid", limits.CPUWeight)
+			}
+			cpuInfo.ControlFlags = winapi.JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | winapi.JOB_OBJECT_CPU_RATE_CONTROL_WEIGHT_BASED
+			cpuInfo.Value = limits.CPUWeight
```

Comment:
I know previously Kevin asked that we avoid having to convert the values we use here from processor relative to job relative, but here you're expecting them to be processor relative but then passing them into `SetInformationObject` as if they were job object relative.

Reply:
Not following this, we just pass in whatever the client passes in directly (unless I'm horribly missing something). We're not doing processing on the user input or anything else

Reply:
Oh I see. I think it's because you're using terms like `ProcessorWeightMin`, etc. Since we're passing the weight given to use directly in, these values are really job object specific. So using limits with names like `JobObjectWeightMin` would make that more clear.